src/core/server/saved_objects/migrationsv2/index.ts

```diff
@@ -0,0 +1,240 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/* eslint-disable no-console */
+
+import { ElasticsearchClient } from 'src/core/server/elasticsearch';
+import { gt } from 'semver';
+import chalk from 'chalk';
+import * as Actions from './actions';
+
+export type ControlState =
+  | 'INIT' // Start
+  | 'DONE' // Migration complete
+  | 'FATAL' // An error occurred, migration didn't succeed
+  | 'INIT_NEW_INDICES' // Blank ES cluster, create new kibana indices
+  | 'SET_SOURCE_WRITE_BLOCK' // Set a write block on the source index to prevent any further writes
+  | 'CLONE_SOURCE'; // Create the target index by cloning the source index
+
+type ActionResponse =
+  | Actions.CloneIndexResponse
+  | Actions.FetchAliasesResponse
+  | Actions.SetIndexWriteBlockResponse;
+// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+export type State = {
+  controlState: ControlState;
+  kibanaVersion: string;
+  source: string;
+  target: string;
+  error?: Error;
+  aliases: Record<string, string>;
+  retryCount: number;
+  retryDelay: number;
+  log: Array<{ level: 'error' | 'info'; message: string }>;
+};
+type Model = (currentState: State, result: ActionResponse) => State;
+type ActionThunk = () => Promise<ActionResponse>;
+type NextAction = (client: ElasticsearchClient, state: State) => ActionThunk | null;
+
+function throwBadControlState(p: never): never;
+function throwBadControlState(controlState: any) {
+  throw new Error('Unknown control state: ' + controlState);
+}
+
+function indexVersion(indexName: string) {
+  return (indexName.match(/\.kibana_(\d+\.\d+\.\d+)_\d+/) || [])[1];
+}
+
+export const model: Model = (currentState: State, res: ActionResponse): State => {
+  const stateP: State = { ...currentState, ...{ log: [] } };
```

Comment:
is it temporary until you add an execution context support?

Reply:
I went back and forth on this a few times. I like that the model is a pure function without any side-effects. So testing doesn't need any mocks and if you inspect the state you have all the information you need on what the behaviour was.

I felt like logging is part of that behaviour that we want to test, it's really important that the right messages are logged. But when you need to pass in a logger mock and inspect the call parameters you loose some of the benefits of the pure model.

But this might be too idealistic and maybe being pragmatic and just using a logger directly is simpler and easier to understand when reading the code for the first time.

Reply:
One thing I like about the pure 'monadic' approach here is that it's harder to accidentally have unintended behavior. Using the logger as an example, it'd be easy to write a test that doesn't check the loggerMock calls at all, but with the pure approach any tests of this form will capture ALL behavior:

```
const newState = model(currentState, res);
expect(newState).toEqual({ ... })
```